# Lab3 - Assignment Sentiment

Copyright: Vrije Universiteit Amsterdam, Faculty of Humanities, CLTL

This notebook describes the LAB-3 assignment of the Text Mining course. It is about sentiment analysis.

The aims of the assignment are:
* Learn how to run a rule-based sentiment analysis module (VADER)
* Learn how to run a machine learning sentiment analysis module (Scikit-Learn/ Naive Bayes)
* Learn how to run scikit-learn metrics for the quantitative evaluation
* Learn how to perform and interpret a quantitative evaluation of the outcomes of the tools (in terms of Precision, Recall, and F<sub>1</sub>)
* Learn how to evaluate the results qualitatively (by examining the data) 
* Get insight into differences between the two applied methods
* Get insight into the effects of using linguistic preprocessing
* Be able to describe differences between the two methods in terms of their results
* Get insight into issues when applying these methods across different  domains

In this assignment, you are going to create your own gold standard set from 50 tweets. You will the VADER and scikit-learn classifiers to these tweets and evaluate the results by using evaluation metrics and inspecting the data.

We recommend you go through the notebooks in the following order:
* **Read the assignment (see below)**
* **Lab3.2-Sentiment-analysis-with-VADER.ipynb**
* **Lab3.3-Sentiment-analysis.with-scikit-learn.ipynb**
* **Answer the questions of the assignment (see below) using the provided notebooks and submit**

In this assignment you are asked to perform both quantitative evaluations and error analyses:
* a quantitative evaluation concerns the scores (Precision, Recall, and F<sub>1</sub>) provided by scikit's classification_report. It includes the scores per category, as well as micro and macro averages. Discuss whether the scores are balanced or not between the different categories (positive, negative, neutral) and between precision and recall. Discuss the shortcomings (if any) of the classifier based on these scores
* an error analysis regarding the misclassifications of the classifier. It involves going through the texts and trying to understand what has gone wrong. It servers to get insight in what could be done to improve the performance of the classifier. Do you observe patterns in misclassifications?  Discuss why these errors are made and propose ways to solve them.

## Credits
The notebooks in this block have been originally created by [Marten Postma](https://martenpostma.github.io) and [Isa Maks](https://research.vu.nl/en/persons/e-maks). Adaptations were made by [Filip Ilievski](http://ilievski.nl).

## Part I: VADER assignments


### Preparation (nothing to submit):
To be able to answer the VADER questions you need to know how the tool works. 
* Read more about the VADER tool in [this blog](https://www.geeksforgeeks.org/python-sentiment-analysis-using-vader/).  
* VADER provides 4 scores (positive, negative, neutral, compound). Be sure to understand what they mean and how they are calculated.
* VADER uses rules to handle linguistic phenomena such as negation and intensification. Be sure to understand which rules are used, how they work, and why they are important.
* VADER makes use of a sentiment lexicon. Have a look at the lexicon. Be sure to understand which information can be found there (lemma?, wordform?, part-of-speech?, polarity value?, word meaning?) What do all scores mean? https://github.com/cjhutto/vaderSentiment/blob/master/vaderSentiment/vader_lexicon.txt) 


### [3.5 points] Question1:

Regard the following sentences and their output as given by VADER. Regard sentences 1 to 7, and explain the outcome **for each sentence**. Take into account both the rules applied by VADER and the lexicon that is used. You will find that some of the results are reasonable, but others are not. Explain what is going wrong or not when correct and incorrect results are produced. 

```
INPUT SENTENCE 1 I love apples
VADER OUTPUT {'neg': 0.0, 'neu': 0.192, 'pos': 0.808, 'compound': 0.6369}

INPUT SENTENCE 2 I don't love apples
VADER OUTPUT {'neg': 0.627, 'neu': 0.373, 'pos': 0.0, 'compound': -0.5216}

INPUT SENTENCE 3 I love apples :-)
VADER OUTPUT {'neg': 0.0, 'neu': 0.133, 'pos': 0.867, 'compound': 0.7579}

INPUT SENTENCE 4 These houses are ruins
VADER OUTPUT {'neg': 0.492, 'neu': 0.508, 'pos': 0.0, 'compound': -0.4404}

INPUT SENTENCE 5 These houses are certainly not considered ruins
VADER OUTPUT {'neg': 0.0, 'neu': 0.51, 'pos': 0.49, 'compound': 0.5867}

INPUT SENTENCE 6 He lies in the chair in the garden
VADER OUTPUT {'neg': 0.286, 'neu': 0.714, 'pos': 0.0, 'compound': -0.4215}

INPUT SENTENCE 7 This house is like any house
VADER OUTPUT {'neg': 0.0, 'neu': 0.667, 'pos': 0.333, 'compound': 0.3612}
```

### Sentence 1: "I love apples" — compound: 0.6369
The result is correct. "Love" has a high positive valence in the VADER lexicon, producing a positive compound score. No special rules are triggered and the output accurately reflects the sentiment.

### Sentence 2: "I don't love apples" — compound: -0.5216
The result is correct. VADER applies its negation rule, detecting "don't" and reversing the valence of "love". This correctly flips the sentiment to negative, showing that VADER handles simple negation well.

### Sentence 3: "I love apples :-)" — compound: 0.7579
The result is correct and stronger than Sentence 1. VADER recognizes ":-)" as a positive emoticon from its built-in emoticon lexicon and adds its valence on top of "love", boosting the compound score above that of Sentence 1.

### Sentence 4: "These houses are ruins" — compound: -0.4404
The result is incorrect. VADER assigns negative valence to "ruins" because the word is associated with destruction in its lexicon. However, here "ruins" is a noun referring to ancient remains, which is neutral. VADER cannot distinguish word senses or part-of-speech, leading to a wrong negative classification.

### Sentence 5: "These houses are certainly not considered ruins" — compound: 0.5867
The result is not necessarily right. The negation of "ruins" flips its valence to positive, and "certainly" acts as an intensifier that amplifies the score. However, the sentence is a neutral descriptive statement. The positive output is an artifact of rule interaction rather than genuine sentiment understanding.

### Sentence 6: "He lies in the chair in the garden" — compound: -0.4215
The result is incorrect. VADER assigns negative valence to "lies" because the word is associated with dishonesty in its lexicon. Here, however, "lies" means "to lie down", which is entirely neutral. This is a clear case of lexical ambiguity that VADER cannot resolve, as it has no mechanism to consider context or word sense.

### Sentence 7: "This house is like any house" — compound: 0.3612
The result is incorrect. VADER treats "like" as a positive sentiment word (as in "to like something") and assigns it positive valence. In this sentence "like" is used as a preposition meaning "similar to", carrying no sentiment. This again illustrates VADER's inability to handle lexical ambiguity.

### [Points: 2.5] Exercise 2: Collecting 50 tweets for evaluation
Collect 50 tweets. Try to find tweets that are interesting for sentiment analysis, e.g., very positive, neutral, and negative tweets. These could be your own tweets (typed in) or collected from the Twitter stream. If you have trouble accessing Twitter, try to find an existing dataset (on websites like kaggle or huggingface).

We will store the tweets in the file **my_tweets.json** (use a text editor to edit).
For each tweet, you should insert:
* sentiment analysis label: negative | neutral | positive (this you determine yourself, this is not done by a computer)
* the text of the tweet
* the Tweet-URL

from:
```
    "1": {
        "sentiment_label": "",
        "text_of_tweet": "",
        "tweet_url": "",
```
to:
```
"1": {
        "sentiment_label": "positive",
        "text_of_tweet": "All across America people chose to get involved, get engaged and stand up. Each of us can make a difference, and all of us ought to try. So go keep changing the world in 2018.",
        "tweet_url" : "https://twitter.com/BarackObama/status/946775615893655552",
    },
```

You can load your tweets with human annotation in the following way.

In [1]:
import json

In [2]:
# import pandas as pd

# df = pd.read_csv('archive(1)/twitter_training.csv', header=None)
# print(df.head())
# print(df.columns)

In [3]:
import pandas as pd
import json

# Load the CSV. Path is relative to this notebook's directory so it works for
# anyone who pulls the repo (the dataset lives in Labs/lab3/archive(1)/).
df = pd.read_csv(
    'archive(1)/twitter_training.csv',
    header=None,
    names=['id', 'entity', 'sentiment', 'tweet_text']
)

# Drop rows with missing tweet text
df = df.dropna(subset=['tweet_text'])


label_map = {
    'Positive': 'positive',
    'Negative': 'negative',
    'Neutral': 'neutral',
    'Irrelevant': 'neutral'  # map irrelevant to neutral
}
df['sentiment_label'] = df['sentiment'].map(label_map)

# try to get a balance: 17 positive, 17 negative, 16 neutral
pos = df[df['sentiment_label'] == 'positive'].sample(17, random_state=42)
neg = df[df['sentiment_label'] == 'negative'].sample(17, random_state=42)
neu = df[df['sentiment_label'] == 'neutral'].sample(16, random_state=42)

sampled = pd.concat([pos, neg, neu]).reset_index(drop=True)

my_tweets = {}
for i, row in sampled.iterrows():
    my_tweets[str(i + 1)] = {
        "sentiment_label": row['sentiment_label'],
        "text_of_tweet": row['tweet_text'], 
        "tweet_url": f"https://twitter.com/i/status/{row['id']}"
    }

# Save to my_tweets.json
with open('my_tweets.json', 'w', encoding='utf-8') as f:
    json.dump(my_tweets, f, indent=4, ensure_ascii=False)

print("Saved 50 tweets to my_tweets.json")

Saved 50 tweets to my_tweets.json


In [4]:
my_tweets = json.load(open('my_tweets.json'))

In [5]:
for id_, tweet_info in my_tweets.items():
    print(id_, tweet_info)
    break

1 {'sentiment_label': 'positive', 'text_of_tweet': 'My birthday consisted of crab legs, watching A Goofy Movie + Lady and the Tramp remake, and now I’m gonna play Overwatch with my best friend!! 🥳🥰', 'tweet_url': 'https://twitter.com/i/status/9309'}


### [5 points] Question 3:

Run VADER on your own tweets (see function **run_vader** from notebook **Lab2-Sentiment-analysis-using-VADER.ipynb**). You can use the code snippet below this explanation as a starting point. 
* [2.5 points] a. Perform a quantitative evaluation. Explain the different scores, and explain which scores are most relevant and why.
* [2.5 points] b. Perform an error analysis: select 10 positive, 10 negative and 10 neutral tweets that are not correctly classified and try to understand why. Refer to the VADER-rules and the VADER-lexicon. Of course, if there are less than 10 errors for a category, you only have to check those. For example, if there are only 5 errors for positive tweets, you just describe those.

In [6]:
def vader_output_to_label(vader_output):
    """
    map vader output e.g.,
    {'neg': 0.0, 'neu': 0.0, 'pos': 1.0, 'compound': 0.4215}
    to one of the following values:
    a) positive float -> 'positive'
    b) 0.0 -> 'neutral'
    c) negative float -> 'negative'
    
    :param dict vader_output: output dict from vader
    
    :rtype: str
    :return: 'negative' | 'neutral' | 'positive'
    """
    compound = vader_output['compound']
    
    if compound < 0:
        return 'negative'
    elif compound == 0.0:
        return 'neutral'
    elif compound > 0.0:
        return 'positive'
    
assert vader_output_to_label( {'neg': 0.0, 'neu': 0.0, 'pos': 1.0, 'compound': 0.0}) == 'neutral'
assert vader_output_to_label( {'neg': 0.0, 'neu': 0.0, 'pos': 1.0, 'compound': 0.01}) == 'positive'
assert vader_output_to_label( {'neg': 0.0, 'neu': 0.0, 'pos': 1.0, 'compound': -0.01}) == 'negative'

In [7]:
# Question 3a
from nltk.sentiment.vader import SentimentIntensityAnalyzer
from sklearn.metrics import classification_report

# Load tweets
with open('my_tweets.json') as f:
    my_tweets = json.load(f)

# Load VADER model
vader_model = SentimentIntensityAnalyzer()

tweets = []
all_vader_output = []
all_vader_labels = []
gold = []

# settings (to change for different experiments)
to_lemmatize = True 
pos = set()

for id_, tweet_info in my_tweets.items():
    the_tweet = tweet_info['text_of_tweet']
    vader_output = vader_model.polarity_scores(the_tweet) # run vader
    vader_label = vader_output_to_label(vader_output) # convert vader output to category
    
    tweets.append(the_tweet)
    all_vader_output.append(vader_output)
    all_vader_labels.append(vader_label)
    gold.append(tweet_info['sentiment_label'])
    
# use scikit-learn's classification report
print(classification_report(gold, all_vader_labels, digits=3))

              precision    recall  f1-score   support

    negative      0.389     0.412     0.400        17
     neutral      0.500     0.188     0.273        16
    positive      0.423     0.647     0.512        17

    accuracy                          0.420        50
   macro avg      0.437     0.415     0.395        50
weighted avg      0.436     0.420     0.397        50



### Question 3a quantitative evaluation

The classification report provides Precision, Recall, and F1-score per class, plus macro and weighted averages.

**Results interpretation:**

- Positive (F1 = 0.512): VADER performs best here. Positive words like excellent, awesome, love are strongly represented in the VADER lexicon and easy to detect. However, the recall score (0.647) significantly exceeds the precision score (0.423), meaning VADER over predicts positive, it flags many neutral and negative tweets as positive.
- Negative (F1 = 0.400): Moderate performance. VADER catches some negative tweets but misses many where the negativity is expressed through context like frustration and sarcasm rather than explicit negative words.
- Neutral (F1 = 0.273): Worst performance. The recall score is very low (0.188), indicating VADER rarely predicts "neutral", it tends to push tweets into positive or negative based on any sentiment word found in the text, even when the tweet as a whole is neutral.
- Macro average (F1 = 0.395): This averages the three class F1 scores equally, regardless of class size. Because our dataset is balanced (17/17/16), this is the most honest single summary of VADER's overall performance. Crucially, the neutral class's poor F1 (0.273) pulls the macro average down just as much as the positive class's relatively better F1 (0.512) pulls it up, nothing is hidden. The result, 0.395, is closer to the worst-performing class than the best, which tells you VADER's performance is uneven rather than consistently moderate.
- Weighted average (F1 = 0.397): This weights each class's F1 by how many examples it has before averaging. Because the dataset is nearly balanced, the weighted average comes out almost identical to the macro average (0.397 vs 0.395).  

The overall accuracy of 42%, which is only marginally above the random chance baseline of 33% for a 3-class problem, indicating VADER provides limited value on this dataset, particularly due to its poor performance on classifying neutral.

**Most Relevant Scores**

The macro F1 (0.395) and per-class F1 are most informative, since the dataset is roughly balanced across three classes. The macro F1 avoids hiding poor neutral performance behind strong positive performance. Accuracy alone is less useful because it equally weights a correct neutral prediction and a correct positive prediction.


In [8]:
# Question 3b
for category in ['positive', 'negative', 'neutral']:
    print(f"{category} — misclassified")
    for i, (tweet, pred, true, scores) in enumerate(zip(tweets, all_vader_labels, gold, all_vader_output)):
        if true == category and pred != true:
            print(f"ID {i+1} | Predicted: {pred} | True: {true} | Compound Score: {scores['compound']}")
            print(f"Tweet: {tweet}")
            print()

positive — misclassified
ID 4 | Predicted: negative | True: positive | Compound Score: -0.1901
Tweet: Procrastination with punk gets a bit ridiculous, but that means more time for Assassins Creed Valhalla.

ID 5 | Predicted: negative | True: positive | Compound Score: -0.1027
Tweet: Did you guys play Herdstein? because I'm obsessed & I feel like I'm cheating on Osrs

ID 6 | Predicted: neutral | True: positive | Compound Score: 0.0
Tweet: That’s right

ID 8 | Predicted: negative | True: positive | Compound Score: -0.7743
Tweet: @PlayStation PS I was with PS for 2 decades. ps3 and r. I will migrate to xbox series x (I already acquired it). was so sad and disappointed in ps5. still have not seen it in physical. and today's presentation October 4. untouchable console! its a joke! goodbye PS

ID 12 | Predicted: negative | True: positive | Compound Score: -0.4404
Tweet: amd gpu’s suck i will remember to switch over nvidia

ID 15 | Predicted: negative | True: positive | Compound Score: -0.186

### Question 3b Error Analysis

#### Positive tweets misclassified

**ID 4** — "Procrastination with punk gets a bit ridiculous, but that means more time for Assassins Creed Valhalla."  
Predicted **negative** (compound: –0.19). VADER picks up "ridiculous" (negative in its lexicon) and ignores the playful, enthusiastic context. The tweet is positively framed, but VADER does not know the context. The word "Ridiculous" here is treated as a straightforwardly negative word.

**ID 5** — "Did you guys play Herdstein? because I'm obsessed & I feel like I'm cheating on Osrs"  
Predicted **negative** (compound: –0.10). The word "cheating" has negative valence in the VADER lexicon. Here it is used figuratively to express enthusiasm, but VADER cannot resolve this figurative usage.

**ID 6** — "That's right"
Predicted **neutral** (compound: 0.0). This is too short for VADER to detect any sentiment, none of its words appear in the lexicon with non-zero valence. The positive label is inferred from conversational context that VADER has no access to.

**ID 8** — "@PlayStation PS I was with PS for 2 decades... was so sad and disappointed in ps5... goodbye PS"
Predicted **negative** (compound: –0.77). The human label is positive (maybe it is reflecting the positive purchase of Xbox), but the tweet contains many strongly negative words like sad and disappointed. The tweet reads negative even to humans without full context. VADER's negative prediction is arguably justified here.

**ID 12** — "amd gpu's suck i will remember to switch over nvidia"
Predicted **negative** (compound: –0.44). VADER correctly detects "suck" as negative. The human annotator labelled this positive, maybe beacuse it might be focusing on the switch to Nvidia, which is positive, but VADER reads the overall tone as negative, which is linguistically defensible.

**ID 15** — "2015 SHOCK OPS IS BATTLEFIELD AT IT'S BEST"
Predicted **negative** (compound: –0.19). The word "shock" has a slight negative valence in the VADER lexicon. The intended meaning is enthusiastic and positive. All-caps should boost intensity per VADER rules, but "shock" itself pulls the score down.

#### Negative tweets misclassified

**ID 20** — "@verizon. I cancelled ALL services. Please restore ALL of my accounts - as of the Right Now!. Like Now"  
Predicted **positive** (compound: +0.65). VADER scores "Right", "Now", and "Please" as positive and polite words. The aggressive, frustrated intent and repetition is not captured. VADER only sees the surface politeness instead of rage.

**ID 26** — "Amazing was how toxic the FIFA community is... Congrats @EASPORTSFIFA for creating one of the six most toxic communities"
Predicted **positive** (compound: +0.86). VADER sees "Amazing" and "Congrats" both strongly positive in its lexicon and misses that they are used sarcastically. VADER has no sarcasm detection. Ironic use of positive words cannot be captured.

**ID 28** — "Tiger tells @GOLFTV: "Stricks is addicted to Fortnite. I don't like Fortnite."" 
Predicted **positive** (compound: +0.36). The explicit negative statement "I don't like Fortnite" should be caught by VADER's negation rule. The word "addicted" kind of like a neutral-to-negative word create competing signals. VADER scores "like" positively and the negation rule is not strong enough to flip the whole compound score negative.

**ID 31** — "It keeps declining my transaction forms... I NEED HELP"  
Predicted **positive** (compound: +0.91). This is a customer complaint, but the tweet includes phrasing like "simply buy", "happy to regularly charge", and the word "help", all of which register as neutral or positive in the VADER lexicon. VADER entirely misses the frustration embedded in the sentence structure.

**ID 32** — "@Blizzard_Ent help me... I NEED TO MUST PLAY WARCRAFT III AT GOD FOR IT!!!!!!!!!!!!!!!"  
Predicted **positive** (compound: +0.82). All-caps and exclamation marks (VADER's intensification rules) amplify the presence of the word "PLAY" and the overall energy. VADER interprets high energy as positive sentiment, while the tweet is an expression of desperate frustration.

#### Neutral tweets misclassified

**ID 36** — "Just kill some things in Call of Duty, modern warfare with viewers"
Predicted **negative** (compound: –0.78). The word "kill" is strongly negative in the VADER lexicon. In a gaming context, "kill" is a technical term with no emotional valence. VADER lacks any domain adaptation for gaming vocabulary.

**ID 40** — "HIGH ELO Best Champions TIER List - League of Legends Patch 10.14"  
Predicted **positive** (compound: +0.82). The words "Best" and "High" are positive in the lexicon. Here they are used as category labels like "Best Champions" and "High Elo" with no evaluative sentiment intended by the author. VADER cannot distinguish descriptive usage from evaluative usage.

**ID 41** — "hell can't say this and it be shit my guy" 
Predicted **negative** (compound: –0.85). The words "hell" and "shit" are strongly negative in the VADER lexicon. This tweet uses internet slang in a neutral and playful way. VADER lacks register and slang awareness.

**ID 45** — "A ban for 2012 Battlefield 4 player Chiiroolaa has occurred"  
Predicted **negative** (compound: –0.87). The word "ban" receives a very negative score in the VADER lexicon. In this context it is a factual announcement. VADER treats any occurrence of "ban" as inherently negative rather than context-dependent.

**ID 49** — "New York charges Johnson & Johnson with insurance fraud over opioid claims"  
Predicted **negative** (compound: –0.71). Word like "fraud" is negative in the lexicon. This is a neutral news headline, just a factual reporting, not an expression of sentiment. VADER is not designed for news headlines, which often contain emotionally charged words in a purely informational context.

**Overall pattern:** 
1. Domain mismatch: words with clear domain-specific neutral meanings like "kill" and "ban" are assigned lexicon-level negative valence.
2. Sarcasm: positive words used sarcastically are not detected as ironic.
3. Context dependent tweets: tweets where sentiment relies on discourse context cannot be analysed by a lexicon-based system.

### [4 points] Question 4:
Run VADER on the set of airline tweets with the following settings:

* Run VADER (as it is) on the set of airline tweets 
* Run VADER on the set of airline tweets after having lemmatized the text
* Run VADER on the set of airline tweets with only adjectives
* Run VADER on the set of airline tweets with only adjectives and after having lemmatized the text
* Run VADER on the set of airline tweets with only nouns
* Run VADER on the set of airline tweets with only nouns and after having lemmatized the text
* Run VADER on the set of airline tweets with only verbs
* Run VADER on the set of airline tweets with only verbs and after having lemmatized the text

* [1 point] a. Generate for all separate experiments the classification report, i.e., Precision, Recall, and F<sub>1</sub> scores per category as well as micro and macro averages. **Use a different code cell (or multiple code cells) for each experiment.**
* [3 points] b. Compare the scores and explain what they tell you.
* - Does lemmatisation help? Explain why or why not.
* - Are all parts of speech equally important for sentiment analysis? Explain why or why not.

In [9]:
# Question 4  Setup
import pathlib
import zipfile
from collections import Counter

import spacy
from sklearn.datasets import load_files
from sklearn.metrics import classification_report
from nltk.sentiment.vader import SentimentIntensityAnalyzer

candidate_paths = [
    pathlib.Path('airlinetweets'),
    pathlib.Path('airlinetweets') / 'airlinetweets',
]
airline_path = next((p for p in candidate_paths if (p / 'positive').exists()), None)
if airline_path is None and pathlib.Path('airlinetweets.zip').exists():
    with zipfile.ZipFile('airlinetweets.zip') as zf:
        zf.extractall('.')
    airline_path = next((p for p in candidate_paths if (p / 'positive').exists()), None)
assert airline_path is not None, 'Could not find the airlinetweets dataset.'

airline_tweets = load_files(
    str(airline_path),
    categories=['negative', 'neutral', 'positive'],
    encoding='utf-8',
    decode_error='replace',
    allowed_extensions=['.txt'],
)
print('Total tweets:', len(airline_tweets.data))
print('Per-class counts:',
      {airline_tweets.target_names[i]: c
       for i, c in Counter(airline_tweets.target).items()})

# Gold labels as strings so they line up with VADER's string predictions
gold_labels = [airline_tweets.target_names[t] for t in airline_tweets.target]

# Models
vader_model = SentimentIntensityAnalyzer()
nlp = spacy.load('en_core_web_sm')


def vader_output_to_label(vader_output):
    """Map VADER compound score to one of {'negative', 'neutral', 'positive'}."""
    compound = vader_output['compound']
    if compound < 0:
        return 'negative'
    elif compound == 0.0:
        return 'neutral'
    else:
        return 'positive'


# spaCy POS tags we care about. spaCy uses universal POS tags:
#   'ADJ' = adjective, 'NOUN' = noun (also 'PROPN' for proper nouns), 'VERB' = verb
POS_FILTERS = {
    'adj':  {'ADJ'},
    'noun': {'NOUN', 'PROPN'},
    'verb': {'VERB'},
}


def preprocess_text(text, pos_filter=None, lemmatize=False):
    """
    Preprocess a tweet according to the experiment settings.

    :param text: raw tweet
    :param pos_filter: None (keep everything) or a set of spaCy POS tags to keep
    :param lemmatize: if True, replace each token with its lemma
    :return: a single string ready to be fed into VADER
    """
    if pos_filter is None and not lemmatize:
       #return raw text
        return text

    doc = nlp(text)
    tokens = []
    for tok in doc:
        if pos_filter is not None and tok.pos_ not in pos_filter:
            continue
        if lemmatize:
            tokens.append(tok.lemma_)
        else:
            tokens.append(tok.text)
    return ' '.join(tokens)


def run_vader_experiment(name, pos_filter=None, lemmatize=False):
    """Run VADER over all airline tweets with the given preprocessing and print the report."""
    predictions = []
    for tweet in airline_tweets.data:
        processed = preprocess_text(tweet, pos_filter=pos_filter, lemmatize=lemmatize)
        scores = vader_model.polarity_scores(processed)
        predictions.append(vader_output_to_label(scores))

    print(f'=== {name} ===')
    print(classification_report(gold_labels, predictions, digits=3, zero_division=0))
    return predictions


Total tweets: 4755
Per-class counts: {'neutral': 1515, 'positive': 1490, 'negative': 1750}


In [10]:
# Experiment 1: VADER as-is (raw text, no preprocessing)
_ = run_vader_experiment('1. VADER as-is')

=== 1. VADER as-is ===
              precision    recall  f1-score   support

    negative      0.795     0.498     0.613      1750
     neutral      0.591     0.527     0.557      1515
    positive      0.562     0.871     0.684      1490

    accuracy                          0.624      4755
   macro avg      0.649     0.632     0.618      4755
weighted avg      0.657     0.624     0.617      4755



In [11]:
# Experiment 2: VADER on lemmatized text
_ = run_vader_experiment('2. VADER + lemmatization', lemmatize=True)

=== 2. VADER + lemmatization ===
              precision    recall  f1-score   support

    negative      0.784     0.520     0.625      1750
     neutral      0.595     0.490     0.537      1515
    positive      0.559     0.881     0.684      1490

    accuracy                          0.623      4755
   macro avg      0.646     0.630     0.615      4755
weighted avg      0.653     0.623     0.616      4755



In [12]:
# Experiment 3: VADER on adjectives only
_ = run_vader_experiment('3. VADER + adjectives only', pos_filter=POS_FILTERS['adj'])

=== 3. VADER + adjectives only ===
              precision    recall  f1-score   support

    negative      0.858     0.204     0.330      1750
     neutral      0.402     0.892     0.554      1515
    positive      0.669     0.438     0.529      1490

    accuracy                          0.496      4755
   macro avg      0.643     0.511     0.471      4755
weighted avg      0.653     0.496     0.464      4755



In [13]:
# Experiment 4: VADER on adjectives only + lemmatized
_ = run_vader_experiment('4. VADER + adjectives + lemmatization',
                         pos_filter=POS_FILTERS['adj'], lemmatize=True)

=== 4. VADER + adjectives + lemmatization ===
              precision    recall  f1-score   support

    negative      0.856     0.205     0.330      1750
     neutral      0.402     0.892     0.554      1515
    positive      0.668     0.438     0.529      1490

    accuracy                          0.497      4755
   macro avg      0.642     0.512     0.471      4755
weighted avg      0.653     0.497     0.464      4755



In [14]:
# Experiment 5: VADER on nouns only
_ = run_vader_experiment('5. VADER + nouns only', pos_filter=POS_FILTERS['noun'])

=== 5. VADER + nouns only ===
              precision    recall  f1-score   support

    negative      0.707     0.169     0.272      1750
     neutral      0.373     0.778     0.504      1515
    positive      0.519     0.411     0.459      1490

    accuracy                          0.439      4755
   macro avg      0.533     0.453     0.412      4755
weighted avg      0.542     0.439     0.405      4755



In [15]:
# Experiment 6: VADER on nouns only + lemmatized
_ = run_vader_experiment('6. VADER + nouns + lemmatization',
                         pos_filter=POS_FILTERS['noun'], lemmatize=True)

=== 6. VADER + nouns + lemmatization ===
              precision    recall  f1-score   support

    negative      0.700     0.182     0.289      1750
     neutral      0.373     0.768     0.502      1515
    positive      0.512     0.405     0.453      1490

    accuracy                          0.439      4755
   macro avg      0.529     0.452     0.414      4755
weighted avg      0.537     0.439     0.408      4755



In [16]:
# Experiment 7: VADER on verbs only
_ = run_vader_experiment('7. VADER + verbs only', pos_filter=POS_FILTERS['verb'])

=== 7. VADER + verbs only ===
              precision    recall  f1-score   support

    negative      0.786     0.286     0.419      1750
     neutral      0.383     0.811     0.520      1515
    positive      0.569     0.350     0.433      1490

    accuracy                          0.473      4755
   macro avg      0.580     0.482     0.458      4755
weighted avg      0.590     0.473     0.456      4755



In [17]:
# Experiment 8: VADER on verbs only + lemmatized
_ = run_vader_experiment('8. VADER + verbs + lemmatization',
                         pos_filter=POS_FILTERS['verb'], lemmatize=True)

=== 8. VADER + verbs + lemmatization ===
              precision    recall  f1-score   support

    negative      0.750     0.291     0.419      1750
     neutral      0.377     0.784     0.509      1515
    positive      0.576     0.358     0.441      1490

    accuracy                          0.469      4755
   macro avg      0.568     0.478     0.457      4755
weighted avg      0.577     0.469     0.455      4755



### Question 4b — Discussion

#### Summary of results

| # | Experiment | macro-F1 | accuracy | neg F1 | neu F1 | pos F1 |
|---|---|---|---|---|---|---|
| 1 | VADER as-is                       | 0.618 | 0.624 | 0.613 | 0.557 | 0.684 |
| 2 | VADER + lemmatized                | 0.615 | 0.623 | 0.625 | 0.537 | 0.684 |
| 3 | adjectives only                   | 0.471 | 0.496 | 0.330 | 0.554 | 0.529 |
| 4 | adjectives only + lemmatized      | 0.471 | 0.497 | 0.330 | 0.554 | 0.529 |
| 5 | nouns only                        | 0.412 | 0.439 | 0.272 | 0.504 | 0.459 |
| 6 | nouns only + lemmatized           | 0.414 | 0.439 | 0.289 | 0.502 | 0.453 |
| 7 | verbs only                        | 0.458 | 0.473 | 0.419 | 0.520 | 0.433 |
| 8 | verbs only + lemmatized           | 0.457 | 0.469 | 0.419 | 0.509 | 0.441 |

The clearest pattern: **VADER on the raw text is the best setting** (macro-F1 = 0.618). Every other configuration is worse. Lemmatisation barely changes anything; POS filtering hurts a lot.

#### Does lemmatisation help?

**The help of lemmatisation is limited.** The macro-F1 drop from VADER as-is (0.618) to VADER + lemmatized (0.615) is only 0.003, which is essentially noise. The same near-zero effect shows up in every paired comparison: experiments 3 vs 4 are identical to three decimals (0.471 vs 0.471), 5 vs 6 differ by 0.002, and 7 vs 8 by 0.001. 

The reason is that VADER's lexicon is built around inflected wordforms found in social media, not lemmas. Words like `loved`, `loves`, and `loving` each have their own entries with their own valence scores. When you lemmatise, all of these collapse to `love`, which is also in the lexicon — so most lookups still succeed and the compound score barely moves. But lemmatisation also strips away things VADER's rule-based system specifically uses: capitalization (`LOVE` triggers an intensification rule), the exact token shape of contractions like `n't`, and emoticons. So lemmatisation either does nothing useful or quietly removes a small amount of signal — and in our results it does both, which cancel out.

Lemmatisation slightly **increased** neg-F1 (0.613 → 0.625) but slightly **decreased** neu-F1 (0.557 → 0.537). That is consistent with the picture above — some negative-sentiment tokens get standardised to lexicon entries, but the structural cues that mark a tweet as truly neutral get washed out.

#### Are all parts of speech equally important for sentiment analysis?

**No — the three POS classes carry very different amounts of sentiment signal.** Restricting VADER to a single POS produces a clear ranking:

**Adjectives carry the most sentiment** (macro-F1 = 0.471). VADER's lexicon is dominated by adjectives — *great, terrible, awful, amazing, rude, helpful, disappointing* — because that is where evaluative meaning lives in English. Airline complaints and praise mostly take the form *"the service was terrible"* or *"the crew was amazing"*, so when you keep only adjectives you still recover a reasonable amount of the sentiment signal.

**Verbs second** (macro-F1 = 0.458). VADER's lexicon does include sentiment-bearing verbs (*love, hate, enjoy, suck, ruin, cancel*), but most verbs in airline tweets are descriptive and neutral (*flew, booked, waited, arrived, called*) and contribute nothing on the sentiment axis. The neg-F1 with verbs only (0.419) is actually higher than with adjectives only (0.330), probably because complaint-specific verbs like *cancelled, delayed, ruined, lost* are highly diagnostic of negative airline tweets — but the overall macro-F1 is still lower than adjectives because positive and neutral classes suffer.

**Nouns carry the least sentiment** (macro-F1 = 0.412). Most nouns in airline tweets are referential — *flight, bag, gate, crew, seat, hour* — and only a handful are inherently evaluative (*delay, nightmare, disaster*). When you filter to nouns only, most tweets contain almost no VADER-lexicon hits, so the compound score collapses to 0 and they all get predicted as neutral. You can see this directly in the numbers: with nouns only, neutral recall jumps to 0.778 (the model is over-predicting neutral) while negative recall crashes to 0.169.


The clearest evidence is what happens to the **negative class** under POS filtering: neg recall drops from 0.498 (as-is) to 0.204 (adj only) and 0.169 (nouns only). Negative tweets in this dataset rely heavily on negation (*"not happy"*, *"won't be flying again"*, *"didn't even apologize"*) and exclamation marks (*"worst flight ever!"*),exactly the tokens that POS filtering throws away. So negative tweets get pushed into the neutral bucket, neutral recall inflates artificially (up to 0.778 with nouns), and the macro average collapses.

#### Bottom line

Lemmatisation is roughly neutral; POS filtering is a meaningful loss. Adjectives > verbs > nouns in terms of how much sentiment signal each POS carries on its own, but no single POS is enough — VADER was designed to operate on the full text, including its punctuation and function words, and any aggressive preprocessing fights against that design.


## Part II: scikit-learn assignments
### [4 points] Question 5
Train the scikit-learn classifier (Naive Bayes) using the airline tweets.

+ Train the model on the airline tweets with 80% training and 20% test set and default settings (TF-IDF representation, min_df=2)
+ Train with different settings:
    + with respect to vectorizing: TF-IDF ('airline_tfidf') vs. Bag of words representation ('airline_count') 
    + with respect to the frequency threshold (min_df). Carry out experiments with increasing values for document frequency (min_df = 2; min_df = 5; min_df =10) 
* [1 point] a. Generate a classification_report for all experiments
* [3 points] b. Look at the results of the experiments with the different settings and try to explain why they differ: 
    + which category performs best, is this the case for any setting?
    + does the frequency threshold affect the scores? Why or why not according to you?

In [18]:
import pathlib
import zipfile
from collections import Counter

import nltk
from nltk.corpus import stopwords
from sklearn.datasets import load_files
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report

# Make sure the NLTK resources we need are available
for find_path, pkg in [
    ('tokenizers/punkt_tab', 'punkt_tab'),
    ('tokenizers/punkt', 'punkt'),
    ('corpora/stopwords', 'stopwords'),
]:
    try:
        nltk.data.find(find_path)
    except LookupError:
        nltk.download(pkg, quiet=True)

# Locate (and if needed, unzip) the airlinetweets dataset.
# The dataset ships as airlinetweets.zip in this folder; the unzipped layout is
#   airlinetweets/{negative,neutral,positive}/*.txt
candidate_paths = [
    pathlib.Path('airlinetweets'),
    pathlib.Path('airlinetweets') / 'airlinetweets',
]
airline_path = next((p for p in candidate_paths if (p / 'positive').exists()), None)
if airline_path is None and pathlib.Path('airlinetweets.zip').exists():
    with zipfile.ZipFile('airlinetweets.zip') as zf:
        zf.extractall('.')
    airline_path = next((p for p in candidate_paths if (p / 'positive').exists()), None)
assert airline_path is not None, 'Could not find the airlinetweets dataset.'
print('Dataset path:', airline_path)

airline_tweets = load_files(
    str(airline_path),
    categories=['negative', 'neutral', 'positive'],
    encoding='utf-8',
    decode_error='replace',
    allowed_extensions=['.txt'],
)
print('Total tweets:', len(airline_tweets.data))
print('Per-class counts:',
      {airline_tweets.target_names[i]: c
       for i, c in Counter(airline_tweets.target).items()})

# 80/20 stratified split, shared across all six experiments so the comparison
# is fair (only the vectoriser/min_df changes between runs).
X_train, X_test, y_train, y_test = train_test_split(
    airline_tweets.data,
    airline_tweets.target,
    test_size=0.20,
    random_state=42,
    stratify=airline_tweets.target,
)


def run_experiment(representation, min_df):
    """Train Multinomial NB with the given setting and print its classification report."""
    vec = CountVectorizer(
        min_df=min_df,
        tokenizer=nltk.word_tokenize,
        stop_words=stopwords.words('english'),
    )
    X_train_counts = vec.fit_transform(X_train)
    X_test_counts = vec.transform(X_test)

    if representation == 'tfidf':
        tfidf = TfidfTransformer()
        X_train_feats = tfidf.fit_transform(X_train_counts)
        X_test_feats = tfidf.transform(X_test_counts)
    else:  # 'count' = Bag of Words
        X_train_feats = X_train_counts
        X_test_feats = X_test_counts

    clf = MultinomialNB().fit(X_train_feats, y_train)
    y_pred = clf.predict(X_test_feats)

    header = f'{representation:>5} | min_df={min_df:<2} | vocab={len(vec.vocabulary_):>5}'
    print('\n=== ' + header + ' ===')
    print(classification_report(
        y_test, y_pred,
        target_names=airline_tweets.target_names,
        digits=3,
        zero_division=0,
    ))


# Six experiments: {airline_tfidf, airline_count} x {min_df=2, 5, 10}
for representation in ('tfidf', 'count'):
    for md in (2, 5, 10):
        run_experiment(representation, md)


Dataset path: airlinetweets
Total tweets: 4755
Per-class counts: {'neutral': 1515, 'positive': 1490, 'negative': 1750}


/opt/anaconda3/envs/CSE412A/lib/python3.10/site-packages/sklearn/feature_extraction/text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/opt/anaconda3/envs/CSE412A/lib/python3.10/site-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ["'d", "'ll", "'m", "'re", "'s", "'ve", 'could', 'might', 'must', "n't", 'need', 'sha', 'wo', 'would'] not in stop_words.
  warnings.warn(



=== tfidf | min_df=2  | vocab= 2480 ===
              precision    recall  f1-score   support

    negative      0.799     0.911     0.852       350
     neutral      0.809     0.686     0.743       303
    positive      0.834     0.826     0.830       298

    accuracy                          0.813       951
   macro avg      0.814     0.808     0.808       951
weighted avg      0.813     0.813     0.810       951



/opt/anaconda3/envs/CSE412A/lib/python3.10/site-packages/sklearn/feature_extraction/text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/opt/anaconda3/envs/CSE412A/lib/python3.10/site-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ["'d", "'ll", "'m", "'re", "'s", "'ve", 'could', 'might', 'must', "n't", 'need', 'sha', 'wo', 'would'] not in stop_words.
  warnings.warn(



=== tfidf | min_df=5  | vocab= 1050 ===
              precision    recall  f1-score   support

    negative      0.805     0.883     0.842       350
     neutral      0.783     0.726     0.753       303
    positive      0.839     0.805     0.822       298

    accuracy                          0.809       951
   macro avg      0.809     0.805     0.806       951
weighted avg      0.809     0.809     0.807       951



/opt/anaconda3/envs/CSE412A/lib/python3.10/site-packages/sklearn/feature_extraction/text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/opt/anaconda3/envs/CSE412A/lib/python3.10/site-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ["'d", "'ll", "'m", "'re", "'s", "'ve", 'could', 'might', 'must', "n't", 'need', 'sha', 'wo', 'would'] not in stop_words.
  warnings.warn(



=== tfidf | min_df=10 | vocab=  568 ===
              precision    recall  f1-score   support

    negative      0.801     0.874     0.836       350
     neutral      0.764     0.746     0.755       303
    positive      0.853     0.782     0.816       298

    accuracy                          0.804       951
   macro avg      0.806     0.801     0.802       951
weighted avg      0.806     0.804     0.804       951



/opt/anaconda3/envs/CSE412A/lib/python3.10/site-packages/sklearn/feature_extraction/text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/opt/anaconda3/envs/CSE412A/lib/python3.10/site-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ["'d", "'ll", "'m", "'re", "'s", "'ve", 'could', 'might', 'must', "n't", 'need', 'sha', 'wo', 'would'] not in stop_words.
  warnings.warn(



=== count | min_df=2  | vocab= 2480 ===
              precision    recall  f1-score   support

    negative      0.824     0.911     0.866       350
     neutral      0.821     0.739     0.778       303
    positive      0.842     0.822     0.832       298

    accuracy                          0.829       951
   macro avg      0.829     0.824     0.825       951
weighted avg      0.829     0.829     0.827       951



/opt/anaconda3/envs/CSE412A/lib/python3.10/site-packages/sklearn/feature_extraction/text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/opt/anaconda3/envs/CSE412A/lib/python3.10/site-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ["'d", "'ll", "'m", "'re", "'s", "'ve", 'could', 'might', 'must', "n't", 'need', 'sha', 'wo', 'would'] not in stop_words.
  warnings.warn(



=== count | min_df=5  | vocab= 1050 ===
              precision    recall  f1-score   support

    negative      0.807     0.897     0.850       350
     neutral      0.801     0.729     0.763       303
    positive      0.843     0.809     0.825       298

    accuracy                          0.816       951
   macro avg      0.817     0.812     0.813       951
weighted avg      0.816     0.816     0.815       951



/opt/anaconda3/envs/CSE412A/lib/python3.10/site-packages/sklearn/feature_extraction/text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/opt/anaconda3/envs/CSE412A/lib/python3.10/site-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ["'d", "'ll", "'m", "'re", "'s", "'ve", 'could', 'might', 'must', "n't", 'need', 'sha', 'wo', 'would'] not in stop_words.
  warnings.warn(



=== count | min_df=10 | vocab=  568 ===
              precision    recall  f1-score   support

    negative      0.803     0.894     0.846       350
     neutral      0.784     0.756     0.770       303
    positive      0.862     0.779     0.818       298

    accuracy                          0.814       951
   macro avg      0.816     0.810     0.811       951
weighted avg      0.815     0.814     0.813       951




=== tfidf | min_df=2  | vocab= 2480 ===
              precision    recall  f1-score   support

    negative      0.799     0.911     0.852       350
     neutral      0.809     0.686     0.743       303
    positive      0.834     0.826     0.830       298

    accuracy                          0.813       951
   macro avg      0.814     0.808     0.808       951
weighted avg      0.813     0.813     0.810       951



/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/feature_extraction/text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ["'d", "'ll", "'m", "'re", "'s", "'ve", 'could', 'might', 'must', "n't", 'need', 'sha', 'wo', 'would'] not in stop_words.
  warnings.warn(



=== tfidf | min_df=5  | vocab= 1050 ===
              precision    recall  f1-score   support

    negative      0.805     0.883     0.842       350
     neutral      0.783     0.726     0.753       303
    positive      0.839     0.805     0.822       298

    accuracy                          0.809       951
   macro avg      0.809     0.805     0.806       951
weighted avg      0.809     0.809     0.807       951



/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/feature_extraction/text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ["'d", "'ll", "'m", "'re", "'s", "'ve", 'could', 'might', 'must', "n't", 'need', 'sha', 'wo', 'would'] not in stop_words.
  warnings.warn(



=== tfidf | min_df=10 | vocab=  568 ===
              precision    recall  f1-score   support

    negative      0.801     0.874     0.836       350
     neutral      0.764     0.746     0.755       303
    positive      0.853     0.782     0.816       298

    accuracy                          0.804       951
   macro avg      0.806     0.801     0.802       951
weighted avg      0.806     0.804     0.804       951



/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/feature_extraction/text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ["'d", "'ll", "'m", "'re", "'s", "'ve", 'could', 'might', 'must', "n't", 'need', 'sha', 'wo', 'would'] not in stop_words.
  warnings.warn(



=== count | min_df=2  | vocab= 2480 ===
              precision    recall  f1-score   support

    negative      0.824     0.911     0.866       350
     neutral      0.821     0.739     0.778       303
    positive      0.842     0.822     0.832       298

    accuracy                          0.829       951
   macro avg      0.829     0.824     0.825       951
weighted avg      0.829     0.829     0.827       951



/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/feature_extraction/text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ["'d", "'ll", "'m", "'re", "'s", "'ve", 'could', 'might', 'must', "n't", 'need', 'sha', 'wo', 'would'] not in stop_words.
  warnings.warn(



=== count | min_df=5  | vocab= 1050 ===
              precision    recall  f1-score   support

    negative      0.807     0.897     0.850       350
     neutral      0.801     0.729     0.763       303
    positive      0.843     0.809     0.825       298

    accuracy                          0.816       951
   macro avg      0.817     0.812     0.813       951
weighted avg      0.816     0.816     0.815       951



/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/feature_extraction/text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ["'d", "'ll", "'m", "'re", "'s", "'ve", 'could', 'might', 'must', "n't", 'need', 'sha', 'wo', 'would'] not in stop_words.
  warnings.warn(



=== count | min_df=10 | vocab=  568 ===
              precision    recall  f1-score   support

    negative      0.803     0.894     0.846       350
     neutral      0.784     0.756     0.770       303
    positive      0.862     0.779     0.818       298

    accuracy                          0.814       951
   macro avg      0.816     0.810     0.811       951
weighted avg      0.815     0.814     0.813       951



### Question 5b — Discussion

Here are the macro-F1 numbers I got across the six runs (80/20 split, `random_state=42`):

| representation | min_df | vocab | macro-F1 | accuracy | neg F1 | neu F1 | pos F1 |
|---|---|---|---|---|---|---|---|
| TF-IDF | 2  | 2480 | 0.808 | 0.813 | 0.852 | 0.743 | 0.830 |
| TF-IDF | 5  | 1050 | 0.806 | 0.809 | 0.842 | 0.753 | 0.822 |
| TF-IDF | 10 |  568 | 0.802 | 0.804 | 0.836 | 0.755 | 0.816 |
| Count  | 2  | 2480 | **0.825** | **0.829** | 0.866 | 0.778 | 0.832 |
| Count  | 5  | 1050 | 0.813 | 0.816 | 0.850 | 0.763 | 0.825 |
| Count  | 10 |  568 | 0.811 | 0.814 | 0.846 | 0.770 | 0.818 |

The pattern is pretty consistent: the **negative** class is the easiest one to predict in every single run. Part of that is just that there are more negative tweets to learn from (around 1750 of them, versus ~1490 positive and ~1515 neutral), so Naive Bayes has a slightly higher prior on "negative" and tends to fall back on it for borderline tweets. But the bigger reason is that complaints have a really obvious vocabulary — words like *delay*, *cancelled*, *hours*, *worst*, *never*, plus the negation token *n't* — and those words barely show up in the other two classes. The model can lean on them.

The hardest class is **neutral**, every time (F1 stays between 0.74 and 0.78). That makes sense if you skim the actual tweets: neutrals are mostly factual questions or follow-ups like "@jetblue DM sent" or "what's the status of flight 1234?", and the words in them (*flight*, airline names, question marks) show up just as often in the other two classes. There's no clean signal for the model to grab onto, so it ends up confusing neutrals with both positives and negatives.

On TF-IDF vs. bag-of-words, BoW wins everywhere, and the biggest gain is on the neutral class (F1 0.74 → 0.78 at `min_df=2`). I think the reason is that TF-IDF down-weights words that appear in lots of tweets, but on this dataset those high-document-frequency words (*thanks*, *delay*, *united*) are exactly the ones that actually carry sentiment. Squashing them hurts more than it helps. It also fits with how Naive Bayes works under the hood — it expects raw counts, so giving it counts directly just feels like the more natural pairing.

The `min_df` threshold matters surprisingly little. Going from 2 → 10 cuts the vocabulary by more than 4× (2480 → 568 features), but accuracy only drops about 1.5 points (0.829 → 0.814 with BoW). Most of the predictive power lives in the high-frequency words, which all three thresholds keep. The drop isn't even across classes though: negative F1 barely moves (0.866 → 0.846), but positive recall takes the biggest hit (0.822 → 0.779) — that's because positives rely on a long tail of compliment words ("amazing", "awesome", "appreciated") that occur only a handful of times each and get filtered out as we raise `min_df`. Neutral F1 actually nudges up a little at `min_df=10`, probably because cutting low-frequency noise makes the model less sure about positive/negative predictions, so more borderline tweets fall into the neutral bucket by default.

So the best of the six setups is **bag-of-words with `min_df=2`** (macro-F1 0.825). The recurring weak spot in all of them is the neutral class, and I don't think any amount of `min_df` tuning is going to fix that — it would need features the bag-of-words model just doesn't have, like negation handling, POS information, or a proper Twitter-aware tokenizer.

### [4 points] Question 6: Inspecting the best scoring features 

+ Train the scikit-learn classifier (Naive Bayes) model with the following settings (airline tweets 80% training and 20% test;  Bag of words representation ('airline_count'), min_df=2)
* [1 point] a. Generate the list of best scoring features per class (see function **important_features_per_class** below) [1 point]
* [3 points] b. Look at the lists and consider the following issues: 
    + [1 point] Which features did you expect for each separate class and why?
    + [1 point] Which features did you not expect and why ? 
    + [1 point] The list contains all kinds of words such as names of airlines, punctuation, numbers and content words (e.g., 'delay' and 'bad'). Which words would you remove or keep when trying to improve the model and why? 

In [19]:
def important_features_per_class(vectorizer, classifier, n=80):
    class_labels = classifier.classes_
    # `get_feature_names()` was removed in recent sklearn versions in favour of
    # `get_feature_names_out()`. Fall back for older installations.
    if hasattr(vectorizer, 'get_feature_names_out'):
        feature_names = vectorizer.get_feature_names_out()
    else:
        feature_names = vectorizer.get_feature_names()
    topn_class1 = sorted(zip(classifier.feature_count_[0], feature_names), reverse=True)[:n]
    topn_class2 = sorted(zip(classifier.feature_count_[1], feature_names), reverse=True)[:n]
    topn_class3 = sorted(zip(classifier.feature_count_[2], feature_names), reverse=True)[:n]
    print("Important words in negative documents")
    for coef, feat in topn_class1:
        print(class_labels[0], coef, feat)
    print("-----------------------------------------")
    print("Important words in neutral documents")
    for coef, feat in topn_class2:
        print(class_labels[1], coef, feat)
    print("-----------------------------------------")
    print("Important words in positive documents")
    for coef, feat in topn_class3:
        print(class_labels[2], coef, feat)

# example of how to call from notebook:
# important_features_per_class(airline_vec, clf)

In [20]:
# Q6 setting: airline tweets, 80/20 split, Bag-of-words ('airline_count'),
# min_df=2, Multinomial Naive Bayes.
# We reuse the X_train / X_test / y_train / y_test split built in Question 5.
airline_vec = CountVectorizer(
    min_df=2,
    tokenizer=nltk.word_tokenize,
    stop_words=stopwords.words('english'),
)
airline_count_train = airline_vec.fit_transform(X_train)
airline_count_test = airline_vec.transform(X_test)

clf = MultinomialNB().fit(airline_count_train, y_train)

# Sanity-check the model before inspecting its features
print(classification_report(
    y_test,
    clf.predict(airline_count_test),
    target_names=airline_tweets.target_names,
    digits=3,
    zero_division=0,
))

# Top-80 features per class (default n=80 as requested in the assignment)
important_features_per_class(airline_vec, clf, n=80)

/opt/anaconda3/envs/CSE412A/lib/python3.10/site-packages/sklearn/feature_extraction/text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/opt/anaconda3/envs/CSE412A/lib/python3.10/site-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ["'d", "'ll", "'m", "'re", "'s", "'ve", 'could', 'might', 'must', "n't", 'need', 'sha', 'wo', 'would'] not in stop_words.
  warnings.warn(


              precision    recall  f1-score   support

    negative      0.824     0.911     0.866       350
     neutral      0.821     0.739     0.778       303
    positive      0.842     0.822     0.832       298

    accuracy                          0.829       951
   macro avg      0.829     0.824     0.825       951
weighted avg      0.829     0.829     0.827       951

Important words in negative documents
0 1513.0 @
0 1388.0 united
0 1198.0 .
0 436.0 ``
0 382.0 flight
0 374.0 ?
0 358.0 !
0 303.0 #
0 208.0 n't
0 156.0 ''
0 127.0 's
0 116.0 service
0 105.0 :
0 102.0 virginamerica
0 94.0 get
0 92.0 customer
0 86.0 plane
0 85.0 cancelled
0 83.0 time
0 83.0 bag
0 82.0 delayed
0 75.0 ...
0 74.0 'm
0 73.0 ;
0 67.0 &
0 66.0 http
0 66.0 hour
0 65.0 hours
0 63.0 late
0 60.0 would
0 60.0 still
0 60.0 airline
0 59.0 -
0 58.0 gate
0 57.0 amp
0 55.0 worst
0 53.0 ca
0 52.0 help
0 52.0 2
0 51.0 waiting
0 51.0 like
0 50.0 flights
0 48.0 one
0 47.0 (
0 47.0 $
0 46.0 've
0 45.0 never
0 45.0 )
0

### Question 6b — Discussion

Looking at the top-80 features the model leans on per class:

The features I *expected* to see were the obvious ones. For the negative class, `n't`, `cancelled`, `service`, `customer`, `bag`, `plane`, `time`, and `flight` all show up near the top, which is basically the vocabulary of an airline complaint — passengers venting about cancellations, lost bags, bad service, and so on. The question mark `?` ranking high on negative also makes sense, since a lot of complaints are phrased as exasperated questions ("why is my flight delayed *again*?"). For positive, the gratitude words come out exactly where I'd expect: `thanks`, `thank`, `great`, `love`, `!`, even `)` (the closing paren from smileys). And for neutral, I expected the dataset to lean on logistical tokens like `flight`, `flights`, `?`, and `http` — tweets asking for status updates or pointing at a DM thread — which is what the list shows.

What I *didn't* expect, or at least didn't want to see, is how much of every class list is taken up by stuff that has nothing to do with sentiment. The airline handles (`united`, `americanair`, `southwestair`, `jetblue`, `usairways`, `virginamerica`) are near the top of all three classes. `united` in particular is the #2 negative feature, and that's not because the word "united" means anything bad — it's just that United gets way more complaints in this dataset than the other airlines, so the model has effectively learned "if it mentions @united, predict negative." That's frequency bias, not sentiment. The same thing is happening with bare punctuation (`@`, `.`, `!`, `?`, `:`, `#`, `''`, `` `` ``, `-`): the NLTK tokenizer splits each of those out as its own token and the default English stop-word list doesn't filter them, so they look "important" just because they're everywhere. And tokens like `http`, `'s`, and `''` are showing up only because of how the tokenizer chops things up, not because they carry meaning.

So if I were trying to improve this model, I'd start by stripping or normalizing the noise. The handles are the single biggest issue — I'd either remove them entirely or replace each one with a generic `<AIRLINE>` placeholder so the model can't shortcut its way to "negative" by spotting "united". I'd also extend the stop-word list to drop the bare punctuation tokens (`.`, `,`, `:`, `''`, `` `` ``, `-`, `#`, `@`), strip URLs with a quick regex (`re.sub(r"https?://\S+|@\w+", "", text)`), and probably drop number/time/flight-number tokens too, since those mostly just push tweets toward neutral without adding any real signal.

The features I'd absolutely keep are the actual content words on each list — *delay*, *cancelled*, *worst*, *bad*, *rude*, *late*, *hours*, *never*, *service*, *customer*, *bag*, *plane* on the negative side, and *thanks*, *thank*, *great*, *love*, *awesome*, *best*, *help* on the positive side. I'd also keep `n't`: negation is one of the strongest sentiment cues a bag-of-words model can possibly use, and the tokenizer is already isolating it nicely. It would be even better to merge it back onto the verb during preprocessing (`did n't` → `didn't` / `did_not`) so the model sees something like "didn't_help" or "didn't_arrive" as a single token. The exclamation mark is the one borderline call — it shows up about 1040 times in positive tweets and only ~360 in negative, so it clearly carries signal, but I'd only keep it if I also added an emoticon-aware tokenizer so things like `:)` and `:-(` get treated the same way.

The big-picture takeaway is that a real chunk of the model's "top features" are dataset artifacts — airline names, punctuation, mentions, URLs — rather than sentiment markers. Cleaning those up, plus handling negation properly, should help the minority classes (positive and neutral) the most, because the model would stop using `united` as a near-synonym for "negative" and would have to actually look at what the tweet is saying.

### [Optional! (will not  be graded)] Question 7
Train the model on airline tweets and test it on your own set of tweets
+ Train the model with the following settings (airline tweets 80% training and 20% test;  Bag of words representation ('airline_count'), min_df=2)
+ Apply the model on your own set of tweets and generate the classification report
* [1 point] a. Carry out a quantitative analysis.
* [1 point] b. Carry out an error analysis on 10 correctly and 10 incorrectly classified tweets and discuss them
* [2 points] c. Compare the results (cf. classification report) with the results obtained by VADER on the same tweets and discuss the differences.

### [Optional! (will not be graded)] Question 8: trying to improve the model
* [2 points] a. Think of some ways to improve the scikit-learn Naive Bayes model by playing with the settings or applying linguistic preprocessing (e.g., by filtering on part-of-speech, or removing punctuation). Do not change the classifier but continue using the Naive Bayes classifier. Explain what the effects might be of these other settings 
+ [1 point] b. Apply the model with at least one new setting (train on the airline tweets using 80% training, 20% test) and generate the scores
* [1 point] c. Discuss whether the model achieved what you expected.

## End of this notebook